### ISSUE TITLE
change the domain API?

### ISSUE THREAD

**Issue by c-cube** [ocaml/ocaml#12385]
Good morning,

This is a bit of a wild one. A big caveat: I'm still quite new to OCaml 5 and actual parallelism in OCaml, so take this with a spoonful of salt.

In the recent past I've discussed with a few people (@gasche, @kayceesrk, @tmattio in particular) about problems I perceive with the concept of `_ Domain.t`. To quote a previous discussion:

---
(begin quoted discussion)

Anyway, something I've been wondering about, and mentionned briefly to Gasche, is that I think that maybe exposing `Domain.t` as a concept is not that great in terms of usability and composability. In a few years, we're going to have programs that use multiple libraries that rely on domains (e.g. domainslib + eio-workpool + some work pool for hashing/compression of tarballs or something like that). However, basing pools on domains is risky because you always need to make sure that `sum_i pool_i.size(I) <= Domain.recommended_count()`; and then it means that each pool is smaller than it needs to be.

So my question/crazy feature suggestion is: given that the Domain API is still unstable, would it be better to instead provide a primitive `spawn_thread_on : int -> (unit -> unit) -> Thread.t` that takes a runtime slice index (from 0 to `recommended_count()-1`), and spawns a thread that uses the corresponding minor heap+runtime struct? The current `'a Domain.t` can be more or less emulated using that[^1], and pools just need to use threads that use disjoint runtime slices. That's basically what I do in moonpool except I have to use domains whose only job is starting threads "on themselves" (to use their own runtime slice).

I think that improves the situation wrt composability, since various pools will still compete for CPU time, but not in a way that degrades performance if there's too many worker threads.

[^1]: a new "fake" domain can be allocated on a runtime slice, picked in a round-robin fashion; and use a little promise/ivar to return an 'a instead of unit.

(end quoted discussion)

---

## summary

So basically my point is: `_ Domain.t` conflates two things that are at odds: 1) a slice of runtime (with its own minor heap), and 2) a thread that does computation and returns a value, running on the given slice of runtime.

But because `Thread.t` already exists, it turns out that the slice of runtime can be used by multiple threads at once, which is more convenient in terms of composability (it's not a big deal to have 40 threads on 8 cores; it's bad™ to have 40 domains on 8 cores, iiuc).

So simply, this proposal is that the function `spawn_thread_on : int -> (unit -> unit) -> Thread.t` that spawns a thread on the runtime with the given index (possibly starting  that runtime if it didn't exist yet) would become the new actual primitive. `Domain.spawn` is just that along with a `int Atomic.t` to round-robin pick the next runtime to use, and should be deprecated. A runtime slice is created with the first thread that uses it, and is deleted once all threads are dead (I suppose this is already the case! There must be refcounting in there I presume).

This is basically what I use in moonpool except I have to run a full `unit Domain.t` for each core to do nothing but spawn threads on itself. It's a bit of a waste of the underying thread.

**Comment by gasche** [comment]
My understanding of the situation is that the design intent of `Domain` is to expose just the bare minimum that the runtime implements, and *not* to provide a good user-facing API for concurrency. 

The base minimum should be nicely documented and all, but users are not encouraged to use it directly. Instead they should be using a concurrency library built on top of Domain, for example one of those you mentioned (Eio: the leviathan of concurrency libraries; Domainslib: a benchmark harness grown organically; moonpool: your own take on this stuff; etc.).

We also hope that the API is flexible enough to implement desired concurrency libraries. If libraries need runtime changes, this is something worth considering. But "user convenience of the Domain module" is a secondary factor.

> [`spawn_thread_on`] is basically what I use in moonpool except I have to run a full unit Domain.t for each core to do nothing but spawn threads on itself. It's a bit of a waste of the underying thread.

I see two separate aspects here:

1. Spawning new domains on demand: you could implement this on the library side by maintaining a mapping from your "runtime slice index" to domains. (A concurrent table of domains currently spawned. When a slice index is passed, if it hasn't been bound a domain yet, use `Domain.spawn` and assign that index to the new domain.) My understanding is that `moonpool` just spawns all the domains on first request, which may be just fine in practice.

2. Having a thread on the domain that only listens for new thread-spawn requests: is it this much of a "waste", can you observe a slowdown in practice? The current runtime already uses a "background thread" for each domain (to service GC interrupts during blocking system calls), and the `Thread` machine also uses a tick thread per domain (to regularly preempt the running thread), so we already have two mostly-idle threads per domain. Does adding such a third thread bring a noticeable cost?

My impression is that doing (1) in the runtime / Domain module would not be easier / more efficient than doing this on the library side. For (2) it may be reasonable easy to tweak the runtime API a bit to support "remotely spawning new threads" in a given domain, instead of requiring a waiter thread on each domain just for this. (My idea: a variant of `caml_c_thread_register` that attaches a thread to an arbitrary domain, instead of just `Dom_c_threads`.) But I would wonder first if we expect actual performance benefits. (Doing this in the runtime risks increase complexity and may create unplanned interactions, so it should come with clear benefits.)

**Comment by c-cube** [comment]
1. spawning/indexing new domains on demand: could be done in the stdlib indeed. I think it is important that it's done in a centralized place because otherwise we still don't get composability.

2. by "waste" I did not really mean CPU overhead. Having an additional thread idling per core is probably not that big a deal but it does mean more processes on the machine (and more noise when looking at the set of threads in gdb, etc.). I suppose it'd be fine to do that for a while and it can be optimized later though. This makes me a bit curious on why the thread ticker, and the domain background thread couldn't be merged? :) (or possibly have only one global ticker that ticks all threads on all alive domains?)


> My understanding of the situation is that the design intent of Domain is to expose just the bare minimum that the runtime implements, and not to provide a good user-facing API for concurrency.

I do understand that, but this is about library writers here. I'm worried that `Domain` as it is is not a great building block _for library writers_, for the reasons above. If Eio wants to have its own domain pool, and Domainslib is also there for heavy CPU tasks, how do you combine the two? Having pools of domains just seems problematic, so I think `Domain` exposes _too much_ in a way, not too little. Exposing only the "start thread on runtime slice" is the more flexible option.

> Instead they should be using a concurrency library built on top of Domain

People will want to use _multiple_ libraries, not pick one, I think. That's what will happen if we want these libraries to be usable as dependencies by other libraries, and that's why  they have to be composable in terms of how many domains/runtime slices they use.

**Comment by kayceesrk** [comment]
I agree that the current domain API is unwieldy for library writers. Specifically, 

> People will want to use multiple libraries, not pick one, I think

I also think that our users should be able to combine libraries freely. 

Today a domain is an expensive resource and spawning more than `Domain.recommended_count()` will degrade performance. There is no easy way for a library to know how many of these fixed resources are actively used and hence libraries are unable to decide how many domains to spawn. 

The proposed
```
spawn_thread_on : int -> (unit -> unit) -> Thread.t
```
also doesn't solve the problem IIUC as it doesn't offer the library writer a way to query whether there is already an active thread at a specific index.

I'd very much like to work on writing down the requirements for the API before proposing a solution.

My requirement are:

1. A domain as a concept should be retained in the runtime system as is. The runtime system makes several assumptions about what domains are and it is much harder to change that wholesale.
2. The proposed solution should let additional domains to be introduced only on demand by the libraries and should also bring down the domains when not used. So any solution that starts off with N domains and maintains them throughout the application lifetime while, say, only one is being actively used by the application is undesirable. The very presence of additional idle domains will make the GC slower due to synchronization overheads.

@polytypic also encountered the same problem earlier and had a proposed library for this: https://polytypic.github.io/idle-domains/idle-domains/Idle_domains/index.html.

**Comment by kayceesrk** [comment]
> @polytypic also encountered the same problem earlier and had a proposed library for this: https://polytypic.github.io/idle-domains/idle-domains/Idle_domains/index.html.

FWIW, I pointed to this library only to say that other library writers have also encountered this problem and not to say that what this library proposes should be the solution.

**Comment by polytypic** [comment]
Apologies for the rambling below. It discusses more than just what immediately relates to the specific proposal here, because in my mind many of these issues are connected &mdash; some more strongly than others. I hope it will give some new ideas to think about around this topic.

Well, first of all, yes, [idle-domains](https://github.com/polytypic/idle-domains) is/was an experiment towards a solution to being able to have various schedulers work more co-operatively. At the point in time when I started experimenting with idle-domains I was under the false impression that threads would be considered a kind of legacy thing. I probably got that impression partly from my incorrect assumption and from starting to read some issue here where the discussions seems to start from that same assumption. Later on in that issue it becomes clear that threads are not going away any time soon.

My thinking on these matters has evolved in the past 6+ months. I spent a good deal of time working on [kcas](https://github.com/ocaml-multicore/kcas/) during those months. A difficulty with Kcas is that it requires [the ability to block](https://github.com/ocaml-multicore/kcas/#blocking-transactions) in a scheduler friendly manner. For that purpose I introduced the [domain-local-await](https://github.com/ocaml-multicore/domain-local-await/) library (incorporating ideas from various people). The point being that instead of e.g. functorizing or parameterizing Kcas with a blocking mechanism (which would make it cumbersome to use) or by e.g. tying Kcas with specific schedulers (think: "kcas-eio" and "kcas-domainslib" and "kcas-moonpool" and ...), the domain-local-await library provides the hinge point or level of indirection.

More recently I created a similar [domain-local-timeout](https://github.com/ocaml-multicore/domain-local-timeout/) library to be able to have scheduler friendly [timeouts](https://github.com/ocaml-multicore/kcas/#timeouts) in Kcas.

These libraries are not only useful for Kcas. It is possible to use them to implement many kinds of concurrent abstractions. Synchronization and communication primitives, IO, and many more things can make use of those. Here is an example of a draft PR of [a lock-free bounded queue](https://github.com/ocaml-multicore/saturn/pull/83) I recently started to develop for the Saturn library.

This line of thinking leads to an idea of ["concurrent runtime services"](https://github.com/polytypic/io#concurrent-runtime-services) that provide minimalistic abstractions of services that concurrent libraries need from the "runtime" (or scheduler). My hypothesis here is that there aren't that many fundamental services that needs to be provided. IOW, that it is possible to come up with a relatively small, manageable, set of services that will support a wide range of higher level things. Note that with "runtime", I'm not talking about the OCaml runtime. I use the term in a more general sense.

The idea with these concurrent runtime services is that by writing higher level libraries (like Kcas, Saturn, IO, ... and others) in terms of those services, the libraries become scheduler independent and can rather be used with any scheduler that provides those services. They can even [work seamlessly across schedulers](https://discuss.ocaml.org/t/interaction-between-eio-and-domainslib-unhandled-exceptions/11971/10?u=polytypic) in a single application.

So, instead of the situation where having two or more schedulers effectively forks the community and every library and application has to choose one or use pervasive parameterization, lots of libraries could actually be scheduler agnostic and would just work with any scheduler that provides the required services. Then a particular application could pick and choose a scheduler or combination of schedulers to run with.

If you followed the link to ["IO should be just a library"](https://github.com/polytypic/io#readme) (or "IO proposal") then there is an idea put forth there that IO should not be tied to a specific scheduler. Rather, IO should be something that can be used from any scheduler (or even outside of any (effects based) scheduler). Currently the IO proposal suggests having a `pollf` abstraction (and little more) for integrating IO with a scheduler. However, today I started thinking that using threads it would actually be possible to do away with the need to explicitly integrate with a scheduler. Asynchronous IO implementation could basically run the necessary "select loop" (by which I mean whatever is needed to interact with a particular low level IO system like io-uring) in its own thread per domain. It would then only need blocking ability (e.g. domain-local-await) from the runtime and would otherwise be independent of any (effects based) schedulers. Calling an IO operation would basically just communicate with that IO thread. (This is also [what the default implementation of domain-local-timeout does](https://github.com/ocaml-multicore/domain-local-timeout/blob/66be5948b6ddf73459b9becd413eea7339960ae6/src/Domain_local_timeout.ml#L65), for example.)

Threads could also work as a mechanism to allow multiple schedulers to share a domain. Due to (semi) pre-emptive scheduling of threads that also gives a kind of fairness to such sharing (which is difficult to achieve otherwise). So, one could, for example, have a queue based scheduler (like Eio) running in a thread on a domain and also share that domain with a work-stealing deque based scheduler (like Domainslib) running in another thread on the same domain (and other domains). The queue based scheduler could service external requests and delegate work to the work-stealing deque based scheduler which would also be able to efficiently move work to other domains.

One question is how expensive threads and switches between threads are in OCaml. This is unclear to me and requires some benchmarking to understand better.

So, let me summarize a bit:

- I think that we really should aim for an ecosystem of mostly compatible and co-operative libraries.

- To do this my main proposal is to provide "concurrent runtime services" that abstract the services needed by concurrent libraries in a scheduler independent manner.

- IO should not be tied to a scheduler. IO can be provided as a scheduler independent library that can then be used from code running on any scheduler that provides blocking.

- Using threads to multiplex domains seems like a promising idea and should be investigated further. Some additional OCaml runtime abilities to control threads might be useful for that purpose.

When it comes to this `spawn_thread_on` proposal in particular. I suspect, like gasche, that it is perhaps not too costly to have a thread for that purpose. You could probably also use that thread for other purposes.

Another thing that comes to mind is that for the purpose of using threads to multiplex domains it might be nice to be able to migrate threads between domains to load balance between domains. So, instead of relying on something like round-robin allocation to make good use of domains, it would actually be possible to move threads from loaded domains to less loaded domains (i.e. something like `migrate: Thread.t -> Domain.id -> unit`). You could even use that to implement a kind of `spawn_thread_on`:

```ocaml
let spawn_thread_on domain_id fn =
  ()
  |> Thread.create @@ fun () ->
     migrate (Thread.self ()) domain;
     fn ()
```

One thing to keep in mind is [possible performance issues on NUMA architectures](https://www.intel.com/content/www/us/en/docs/vtune-profiler/cookbook/2023-0/os-thread-migration.html). So, being able to directly start a thread on a specific domain might make sense from that POV.

Also, on a higher level, yes, assuming one would use threads to multiplex domains, then having some sort of "domain - thread" manager makes sense. I don't have an immediate proposal for that (I have some ideas), but I think that `spawn_thread_on` is probably not quite sufficient for that.

**Comment by c-cube** [comment]
> The proposed
>
> spawn_thread_on : int -> (unit -> unit) -> Thread.t 
>
> also doesn't solve the problem IIUC as it doesn't offer the library writer a way to query whether there is already an active thread at a specific index.

I didn't reply to that earlier. The point is that it's perfectly fine to have multiple threads in one domain, they won't degrade performance the same way domains do afaict. The OS scheduler (and OCaml's ticker thread?) will make sure they are preempted regularly. 

There could be a count of how many threads are at a given index though (useful to know when to retire the domain, when this count reaches 0). Then one can put the new thread in a less busy domain index. 

One of my tests in moonpool is to take a heavy computation that runs on a (thread) pool, and run multiple of this computation in parallel (thus creating more threads than there are cores). This could be a basis for benchmarking composable approaches: measuring the overhead of running in parallel multiple computations that each expect to use all cores already.

**Comment by kayceesrk** [comment]
> There could be a count of how many threads are at a given index though (useful to know when to retire the domain, when this count reaches 0). Then one can put the new thread in a less busy domain index.

My point was specifically this. Without a way to know whether a domain already has a thread running on it, two different libraries may both start creating threads from domain index 0. If there were 4 cores available, and each library wants to run 2 threads in parallel, then ideally each thread should be mapped uniquely to a core. If one naively starts at 0 for both the libraries, then libraries will end up spawning their threads on the domains at index 0 and 1, underutilising the available parallelism. 

I did wonder whether the API should itself not take care of spawning the thread on a free domain if available, creating domains up to the limit of `Domain.recommended_count`? If all domains are occupied, then we can start oversubscribing the domains with the new threads.

**Comment by c-cube** [comment]
Yes, you are right. I would couple this with an int atomic to pick the next domain index in a round Robin fashion.

Generally I think the best we can hope, in terms of compositionality, is to have each library get guarantees about parallelism within the threads it allocates itself. So for domainslib you can build a work pool and be sure that each thread is on a different domain.

However these threads might have to share with threads from other pools. I don't see a good way of avoiding that entirely.

**Comment by gasche** [comment]
(My comment here had been placed in the wrong issue, apologies and thanks @dbuenzli for the notification.)

**Comment by lpw25** [comment]
@gasche I think you meant to post that on #12368

**Comment by kayceesrk** [comment]
> I would couple this with an int atomic to pick the next domain index in a round Robin fashion.

We may be able to do a little better than round robin. We could maintain a stack of next domain ids, initially loaded with `0` to `Domain.recommended_count() - 1`. Spawning threads can pop one if available and spawn the thread there. If not, pick a random id. This need not be particularly fast and we can use a lock to protect the stack.

**Comment by c-cube** [comment]
Ok, I like that (assuming that domains can wind down when there's no thread on them, it's indeed more flexible.)
